File: src/main/kotlin/day01/solution.kt
package day01

import java.io.File
import kotlin.math.abs

fun parse(inp: String) = inp.lineSequence()
    .map { line -> line.split("   ").map { it.toInt() }.let { it[0] to it[1] } }
    .unzip()

fun first(inp: String): Int {
    val (left, right) = parse(inp)
    return left.sorted()
        .zip(right.sorted()) { a, b -> abs(a - b) }
        .sum()
}

fun second(inp: String): Int {
    val (left, right) = parse(inp)
    return left.sumOf { num -> num * right.count { it == num } }
}

fun main() {
    val testInput = File("src/main/kotlin/day01/test-input.txt").readText()
    val input = File("src/main/kotlin/day01/input.txt").readText()
    println(first(input))
    println(second(input))
}

File: haskell/day01/solution.hs
import Data.List (sort)
import Data.List.Split (splitOn)

toTuple :: [Int] -> (Int, Int)
toTuple [x, y] = (x, y)

parse :: String -> ([Int], [Int])
parse inp = unzip $ map (toTuple . map read . splitOn "   ") (lines inp)

first :: String -> Int
first inp =
  let (left, right) = parse inp
   in sum $ zipWith (\a b -> abs $ a - b) (sort left) (sort right)

second :: String -> Int
second inp =
  let (left, right) = parse inp
   in sum $ map (\x -> x * length (filter (== x) right)) left

main :: IO ()
main = do
  testInput <- readFile "../../src/main/resources/day01/test-input.txt"
  input <- readFile "../../src/main/resources/day01/input.txt"
  print $ first input
  print $ second input
